### PR TITLE
[IEDriver] fix Edge IE mode issue that sometimes cannot get the handle of a newly opened window

### LINKED ISSUE #19: IWebDriver.WindowHandles in Selenium 4 does not return correct number of open browser windows
I updated Selenium to 4. I test legacy app that runs only on IE. Now IE will be closed down and I need to switch to MS Egde in capability mode. My tests during execution need to switch to new browser windows and uses function IWebDriver.WindowHandles to do it (code below).

Comparing to Selenium 3 I had to add new options for a driver to run test on Edge like this:

var options = new InternetExplorerOptions();
options.AttachToEdgeChrome = true;
options.EdgeExecutablePath = @"\msedge.exe";

Then I use code like this to switch between browser windows when user open new browser window in tested application:

public Context SwitchToNewIEWindow(string windowTitle, bool isEqual = true)
{
for (int i = 0; i < context.GetTimeout().Seconds; i++)
{
try
{
var ieWindowHandlers = context.GetDriver().WindowHandles;
foreach(var handler in ieWindowHandlers)
{
var window = context.GetDriver().SwitchTo().Window(handler);
bool condition = isEqual ? window.Title == windowTitle : window.Title.Contains(windowTitle);
if (condition)
{
context.SetDriver(window);
context.SetCurrentWindow(windowTitle);
return context;
}
}
}
catch (Exception)
{
context.Wait(1);
}
}
throw new Exception($"No window found with title: {windowTitle}");

    }

In Selenium 3 everything works fine and WindowHandles function returns 2 handles. In Selenium 4 and IEDriver in Edge mode this function returns only 1 handle of main browser window. In my case waiting a few seconds ( as it is recommended in guidelines) does not change anything. Even after a minute I still get only handle of first window. Any suggestions how to make it work properly again in Selenium 4. 

I am aware of issue https://github.com/SeleniumHQ/selenium/issues/8868 but I was redirected here as a correct place to report this issue.

### LINKED ISSUE #8868: In Edge(IE mode), I can't get the handle of window which opened by window.open
## 🐛 Bug Report

In Edge(IE mode), I can't get the handle of window which opened by window.open.
It's no problem in IE.

## To Reproduce

1. Use Edge in IE mode
https://docs.microsoft.com/en-us/deployedge/edge-ie-mode-policies

2. Follow the directions for the “Required Configuration”
https://github.com/SeleniumHQ/selenium/wiki/InternetExplorerDriver

3. In html page, make a button to open a new window

4. In test script, use Edge in IE mode by the following code
ieOptions = webdriver.IeOptions()
ieOptions.add_additional_option("ie.edgechromium", True)
ieOptions.add_additional_option("ie.edgepath",'C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe')
driver = webdriver.Ie(executable_path='./driver_machine/IEDriverServer.exe', options=ieOptions)

5. In test script, click the button to open a new window

6. In test script, get the handle of window by driver.window_handles
  
the result is 
1. before click driver.window_handles:['3d24e02d-2013-44d6-bf64-c8800d999903']
2. after click driver.window_handles:['3d24e02d-2013-44d6-bf64-c8800d999903']

only get the handle of old window, can't get the handle of new window

## Expected behavior

in the step 6, two handle of old and new window can be geted

In IE the result is OK. 
1. before click driver.window_handles:['730db730-41be-45c5-9bed-1491f3ac0f8d']
2. after click driver.window_handles:['730db730-41be-45c5-9bed-1491f3ac0f8d','4bc6f24a-f672-4bfe-9919-fd343410b58a']

## Test script or set of commands reproducing this issue

test.py:
from selenium import webdriver
import time

#Edge(IE mode)
ieOptions = webdriver.IeOptions()
ieOptions.add_additional_option("ie.edgechromium", True)
ieOptions.add_additional_option("ie.edgepath",'C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe')
driver = webdriver.Ie(executable_path='./driver_machine/IEDriverServer.exe', options=ieOptions)
#IE
#driver = webdriver.Ie(executable_path='./driver_machine/IEDriverServer.exe')

driver.implicitly_wait(30)
driver.maximize_window()
driver.get('http://localhost:8021/test/test1.html')
time.sleep(5)
print("1. before click driver.window_handles:" + str(driver.window_handles))

_btn1 = driver.find_element_by_id('_btn1')
_btn1.click()
time.sleep(5)
print("2. after click driver.window_handles:" + str(driver.window_handles))

test1.html:
&lt;!DOCTYPE html&gt;
&lt;html&gt;
&lt;head&gt;
    &lt;title&gt;test1&lt;/title&gt;
    &lt;script type=&quot;text/javascript&quot;&gt;
    function testWindowOpen() {
        window.open(&#x27;test2.html&#x27;, &#x27;test2window&#x27;, &#x27;width=800px, height=600px, menubar=no,toolbar=no,status=no,location=no,resizable=yes,scrollbars=no,titlebar=no&#x27;);
    }
    &lt;/script&gt;
&lt;/head&gt;
&lt;body&gt;
    &lt;input id=&quot;_btn1&quot; type=&quot;button&quot; value=&quot;window.open&quot; onclick=&quot;testWindowOpen();&quot; /&gt;
&lt;/body&gt;
&lt;/html&gt;

test2.html:
&lt;!DOCTYPE html&gt;
&lt;html&gt;
&lt;head&gt;
    &lt;title&gt;test2&lt;/title&gt;
&lt;/head&gt;
&lt;body&gt;
    test2
&lt;/body&gt;
&lt;/html&gt;

## Environment

OS: Windows 10 2004
Browser: Edge IE11
Browser version: Edge(85.0.564.51) IE11(11.329.19041.0)
Browser Driver version: IEDriverServer 3.150.1
Language Bindings version: Python 3.141.0 
Selenium Grid version (if applicable):

### PR DESCRIPTION
**Pull request by yusuke-noda**
### Description

Fix Edge IE mode issue that sometimes cannot  get the handle of the newly opened window.

### Motivation and Context

This is a fix for an issue of Edge IE mode that sometimes cannot get the handle of a newly opened window.
Resolve #8868
Resolve MicrosoftEdge/EdgeWebDriver#19

### Reproduction steps
Attached zip file contains test programs to reproduce of the issue.
[iedrivertest.zip](https://github.com/SeleniumHQ/selenium/files/8784854/iedrivertest.zip)

1.  Install node_modules
`npm install`
2. Run server.
`node server.js`
3. Run tests several times.
`node test.js --edge-path "path\to\msedge.exe" --driver-path "path\to\IEDriverServer.exe"`
`node test2.js --edge-path "path\to\msedge.exe" --driver-path "path\to\IEDriverServer.exe"`
`node test3.js --edge-path "path\to\msedge.exe" --driver-path "path\to\IEDriverServer.exe"`
`node test4.js --edge-path "path\to\msedge.exe" --driver-path "path\to\IEDriverServer.exe"`
`node test5.js --edge-path "path\to\msedge.exe" --driver-path "path\to\IEDriverServer.exe"`

#### Contents of test programs
1. test.js
Open one child window and check that can get window handle of the child window.
2. test2.js
Open four child window and check that can get window handles of all of the child windows.
3. test3.js
Open four child tabs in the same window that contains the parent tab and check that can get window handles of all of the child tabs.
4. test4.js
Open child window and its grandchild window at same time and close child window immediately, then check that can get window handle of grandchild windows.
5. test5.js
Open child window and its grandchild window at same time, then check that can get window handle of child and grandchild windows.

### Unresolved problem
Open child and its grandchild tabs in the same window that contains the parent tab at same time, IEDriver cannot get window handle of the child tab.
`node test6.js --edge-path "path\to\msedge.exe" --driver-path "path\to\IEDriverServer.exe"`

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by ven92** [comment]
I'm facing the same issue. I tried with IE driver v4.0 & v4.2. MS Edge is getting launched in IE Mode, window handle works when only one Window present. But if multiple window gets open, then the code get stuck at driver.window_handles line. Could you please provide the ASAP..

**Comment by JeyBee83** [comment]
Any updates on this? A merge would be very helpful, we have lots of issues with this.

**Comment by bwalderman** [comment]
When I built and tested IEDriverServer this change, I found that test.js, test2.js and test3.js are more stable, but test4.js and test5.js are still failing. @yusuke-noda does this match your expectations?

Log from test4.js:

```
> node .\test4.js --driverPath=IEDriverServer.exe --edgePath="C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"
start
iteration start 0
retry
retry
retry
Error: too many retry
    at C:\Users\brwalder\Desktop\iedrivertest\test4.js:51:17
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
finally
```

Log from test5.js:

```
> node .\test5.js --driverPath=IEDriverServer.exe --edgePath="C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"
start
iteration start 0
retry
retry
retry
Error: too many retry
    at C:\Users\brwalder\Desktop\iedrivertest\test5.js:51:17
    at processTicksAndRejections (node:internal/process/task_queues:94:5)
finally
```

**Comment by yusuke-noda** [comment]
Hello @bwalderman .
On my environments (Windows 10 21H2 and Windows 11), test4.js and test5.js are also stable.

```
node .\test4.js --driver-path .\IEDriverServer.exe --edge-path "C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"
start
iteration start 0
iteration finished 0
iteration start 1
iteration finished 1
iteration start 2
iteration finished 2
iteration start 3
iteration finished 3
iteration start 4
iteration finished 4
iteration start 5
iteration finished 5
iteration start 6
iteration finished 6
iteration start 7
iteration finished 7
iteration start 8
iteration finished 8
iteration start 9
iteration finished 9
iteration start 10
iteration finished 10
iteration start 11
iteration finished 11
iteration start 12
iteration finished 12
iteration start 13
iteration finished 13
iteration start 14
iteration finished 14
iteration start 15
iteration finished 15
iteration start 16
iteration finished 16
iteration start 17
iteration finished 17
iteration start 18
iteration finished 18
iteration start 19
iteration finished 19
finally
```

```
node .\test5.js --driver-path .\IEDriverServer.exe --edge-path "C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe"
start
iteration start 0
[ 'child51', 'child52' ]
iteration finished 0
iteration start 1
[ 'child51', 'child52' ]
iteration finished 1
iteration start 2
[ 'child51', 'child52' ]
iteration finished 2
iteration start 3
[ 'child51', 'child52' ]
iteration finished 3
iteration start 4
[ 'child51', 'child52' ]
iteration finished 4
iteration start 5
[ 'child51', 'child52' ]
iteration finished 5
iteration start 6
[ 'child51', 'child52' ]
iteration finished 6
iteration start 7
[ 'child51', 'child52' ]
iteration finished 7
iteration start 8
[ 'child51', 'child52' ]
iteration finished 8
iteration start 9
[ 'child51', 'child52' ]
iteration finished 9
iteration start 10
[ 'child51', 'child52' ]
iteration finished 10
iteration start 11
[ 'child51', 'child52' ]
iteration finished 11
iteration start 12
[ 'child51', 'child52' ]
iteration finished 12
iteration start 13
[ 'child51', 'child52' ]
iteration finished 13
iteration start 14
[ 'child51', 'child52' ]
iteration finished 14
iteration start 15
[ 'child51', 'child52' ]
iteration finished 15
iteration start 16
[ 'child51', 'child52' ]
iteration finished 16
iteration start 17
[ 'child51', 'child52' ]
iteration finished 17
iteration start 18
[ 'child51', 'child52' ]
iteration finished 18
iteration start 19
[ 'child51', 'child52' ]
iteration finished 19
finally
```

But if "Always open pop-ups in a new tab" option is enabled in IE's tabbed Browsing Settings, test5.js fails same as test6.js

**Comment by Oleksii-tmnas** [comment]
hey guys any updates on this? looking forward to the changes to be merged...

**Comment by bwalderman** [comment]
The "Always open pop-ups in a new tab" setting isn't visible in my Windows 11 environment anymore, probably because of the IE11 removal. So I can't verify that test5.js and test6.js would be fixed. However, the change looks good to me and the other test cases appear to be stable.

I've just left a couple suggestions to rename variables to match the style used in the rest of the code.

File: cpp/iedriver/IECommandExecutor.cpp [hunk 1]
```diff
@@ -305,6 +305,9 @@ LRESULT IECommandExecutor::OnAfterNewWindow(UINT uMsg,
       // sleep 0.5s then get current window handles
       clock_t end = clock() + (DEFAULT_BROWSER_REATTACH_TIMEOUT_IN_MILLISECONDS / 1000 * CLOCKS_PER_SEC);
       std::vector<HWND> diff;
+      int loopCount = 0;
```
Comment (bwalderman) [inline cpp/iedriver/IECommandExecutor.cpp]:
```suggestion
      int loop_count = 0;
```

File: cpp/iedriver/IECommandExecutor.cpp [hunk 2]
```diff
@@ -305,6 +305,9 @@ LRESULT IECommandExecutor::OnAfterNewWindow(UINT uMsg,
...
+      bool isProcessWM = false;
```
Comment (bwalderman) [inline cpp/iedriver/IECommandExecutor.cpp]:
```suggestion
      bool is_processing_wm = false;
```

File: cpp/iedriver/IECommandExecutor.cpp [hunk 3]
```diff
@@ -330,38 +333,55 @@ LRESULT IECommandExecutor::OnAfterNewWindow(UINT uMsg,
...
+              ::Sleep(100);
+              is_ready = this->factory_->IsBrowserProcessInitialized(process_id);
+            }
+
+            ProcessWindowInfo info;
+            info.dwProcessId = process_id;
+            info.hwndBrowser = new_window_window;
+            info.pBrowser = NULL;
+            std::string error_message = "";
+            bool attachFlag = this->factory_->AttachToBrowser(&info, &error_message);
```
Comment (bwalderman) [inline cpp/iedriver/IECommandExecutor.cpp]:
```suggestion
            bool attach_flag = this->factory_->AttachToBrowser(&info, &error_message);
```

**Review by bwalderman — APPROVED** [review]

File: cpp/iedriver/IECommandExecutor.cpp (hunk 3 above)
Comment (yusuke-noda) [reply cpp/iedriver/IECommandExecutor.cpp]: thank you for your suggestions.
I've renamed variables.

**Review by diemol — APPROVED** [review]
Thanks for reviewing, @bwalderman!

Thank you for this PR, @yusuke-noda!

**Comment by titusfortner** [comment]
IE Driver 4.3 is released - https://github.com/SeleniumHQ/selenium/releases/tag/selenium-4.3.0

**Comment by anishamarotkar** [comment]
With Selenium Java version 4.4.0 or 4.3.0 and IE Driver 4.3.0 I am still getting facing the same issue. Can anyone please help?

**Comment by jyosha3** [comment]
getting the same issue with latest versions of  IE driver and selenium java versions.Any one was able to find a solution or workaround?

**Comment by titusfortner** [comment]
@anishamarotkar & @jyosha3 can you comment on https://github.com/MicrosoftEdge/EdgeWebDriver/issues/19 we may need to re-open that.

**Comment by jyosha3** [comment]
@titusfortner Yes did comment on that issue just now.How can it be reopened though.I am new and came here looking for a solution so dont know how things work here.

**Comment by titusfortner** [comment]
Yeah, since IEDriver is *only supported via Edge right now, the code management is being done in collaboration with Microsoft. So the current practice (subject to change) is to file the issue in MicrosoftEdge repo, and then the PRs are done here. @bwalderman is the person who would be able to re-open that issue. Also, you might need to provide more info with a trace log in a reproducible case if it is at all different from what has been previously reported.

**Comment by jyosha3** [comment]
@titusfortner ok .. thanks for the information.. i have provided a test scenario in the microsoft edge repo.. alongwith the exception i gt. @bwalderman  do I need to provide any more details for u to reopen / address the issue.

**Comment by tenisaname** [comment]
Hello, are there any changes on this issue?
